### ISSUE TITLE
Define entities in kotlin abstract classes?

### ISSUE THREAD

**Issue by polmabri** [requery/requery#468]
Hey,

is it possible to use kotlin abstract classes to define entities? I tried it and got an error.

works:

```kotlin
@Entity
interface MyEntity : Persistable {

    @get:Key
    @get:Generated
    val id: Int
}
```

generated entity class for the interface:

```kotlin
public class MyEntityEntity implements MyEntity, Persistable {
    public static final AttributeDelegate<MyEntityEntity, Integer> ID = new AttributeDelegate(
    new AttributeBuilder<MyEntityEntity, Integer>("id", int.class)
    .setProperty(new IntProperty<MyEntityEntity>() {
        @Override
        public Integer get(MyEntityEntity entity) {
            return entity.id;
        }

        @Override
        public void set(MyEntityEntity entity, Integer value) {
            entity.id = value;
        }

        @Override
        public int getInt(MyEntityEntity entity) {
            return entity.id;
        }

        @Override
        public void setInt(MyEntityEntity entity, int value) {
            entity.id = value;
        }
    })
    .setPropertyName("getId")
    .setPropertyState(new Property<MyEntityEntity, PropertyState>() {
        @Override
        public PropertyState get(MyEntityEntity entity) {
            return entity.$id_state;
        }

        @Override
        public void set(MyEntityEntity entity, PropertyState value) {
            entity.$id_state = value;
        }
    })
    .setKey(true)
    .setGenerated(true)
    .setLazy(false)
    .setNullable(false)
    .setUnique(false)
    .build());

    public static final Type<MyEntityEntity> $TYPE = new TypeBuilder<MyEntityEntity>(MyEntityEntity.class, "MyEntity")
    .setBaseType(MyEntity.class)
    .setCacheable(true)
    .setImmutable(false)
    .setReadOnly(false)
    .setStateless(false)
    .setView(false)
    .setFactory(new Supplier<MyEntityEntity>() {
        @Override
        public MyEntityEntity get() {
            return new MyEntityEntity();
        }
    })
    .setProxyProvider(new Function<MyEntityEntity, EntityProxy<MyEntityEntity>>() {
        @Override
        public EntityProxy<MyEntityEntity> apply(MyEntityEntity entity) {
            return entity.$proxy;
        }
    })
    .addAttribute(ID)
    .build();

    private PropertyState $id_state;

    private int id;

    private final transient EntityProxy<MyEntityEntity> $proxy = new EntityProxy<MyEntityEntity>(this, $TYPE);

    public MyEntityEntity() {
    }

    @Override
    public int getId() {
        return $proxy.get(ID);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof MyEntityEntity && ((MyEntityEntity)obj).$proxy.equals(this.$proxy);
    }

    @Override
    public int hashCode() {
        return $proxy.hashCode();
    }

    @Override
    public String toString() {
        return $proxy.toString();
    }
}
```

throws "cannot find symbol variable id" error:

```kotlin
@Entity
abstract class MyEntity : Persistable {

    @get:Key
    @get:Generated
    abstract val id: Int
}
```

generated entity for the abstract class:
```kotlin
public class MyEntityEntity extends MyEntity implements Persistable {
    public static final AttributeDelegate<MyEntityEntity, Integer> ID = new AttributeDelegate(
    new AttributeBuilder<MyEntityEntity, Integer>("id", int.class)
    .setProperty(new IntProperty<MyEntityEntity>() {
        @Override
        public Integer get(MyEntityEntity entity) {
            return entity.id;
        }

        @Override
        public void set(MyEntityEntity entity, Integer value) {
            entity.id = value;
        }

        @Override
        public int getInt(MyEntityEntity entity) {
            return entity.id;
        }

        @Override
        public void setInt(MyEntityEntity entity, int value) {
            entity.id = value;
        }
    })
    .setPropertyName("getId")
    .setPropertyState(new Property<MyEntityEntity, PropertyState>() {
        @Override
        public PropertyState get(MyEntityEntity entity) {
            return entity.$id_state;
        }

        @Override
        public void set(MyEntityEntity entity, PropertyState value) {
            entity.$id_state = value;
        }
    })
    .setKey(true)
    .setGenerated(true)
    .setLazy(false)
    .setNullable(false)
    .setUnique(false)
    .build());

    public static final Type<MyEntityEntity> $TYPE = new TypeBuilder<MyEntityEntity>(MyEntityEntity.class, "MyEntity")
    .setBaseType(MyEntity.class)
    .setCacheable(true)
    .setImmutable(false)
    .setReadOnly(false)
    .setStateless(false)
    .setView(false)
    .setFactory(new Supplier<MyEntityEntity>() {
        @Override
        public MyEntityEntity get() {
            return new MyEntityEntity();
        }
    })
    .setProxyProvider(new Function<MyEntityEntity, EntityProxy<MyEntityEntity>>() {
        @Override
        public EntityProxy<MyEntityEntity> apply(MyEntityEntity entity) {
            return entity.$proxy;
        }
    })
    .addAttribute(ID)
    .build();

    private PropertyState $id_state;

    private final transient EntityProxy<MyEntityEntity> $proxy = new EntityProxy<MyEntityEntity>(this, $TYPE);

    public MyEntityEntity() {
    }

    @Override
    public int getId() {
        return $proxy.get(ID);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof MyEntityEntity && ((MyEntityEntity)obj).$proxy.equals(this.$proxy);
    }

    @Override
    public int hashCode() {
        return $proxy.hashCode();
    }

    @Override
    public String toString() {
        return $proxy.toString();
    }
}
```

the generated entity class code is nearly the same, but ``` private int id;``` property is missing in the entity class generated for the abstract class.

**Comment by consp1racy** [comment]
And if `val id` is not `abstract` but `open`?

**Comment by polmabri** [comment]
```kotlin
@Entity
abstract class MyEntity : Persistable {

    @get:Key
    @get:Generated
    open val id: Int? = null
}
```

results in "id has private access in MyEntity" error for entity class:

```kotlin
public class MyEntityEntity extends MyEntity implements Persistable {
    public static final AttributeDelegate<MyEntityEntity, Integer> ID = new AttributeDelegate(
    new AttributeBuilder<MyEntityEntity, Integer>("id", Integer.class)
    .setProperty(new Property<MyEntityEntity, Integer>() {
        @Override
        public Integer get(MyEntityEntity entity) {
            return entity.id;
        }

        @Override
        public void set(MyEntityEntity entity, Integer value) {
            entity.id = value;
        }
    })
    .setPropertyName("getId")
    .setPropertyState(new Property<MyEntityEntity, PropertyState>() {
        @Override
        public PropertyState get(MyEntityEntity entity) {
            return entity.$id_state;
        }

        @Override
        public void set(MyEntityEntity entity, PropertyState value) {
            entity.$id_state = value;
        }
    })
    .setKey(true)
    .setGenerated(true)
    .setLazy(false)
    .setNullable(true)
    .setUnique(false)
    .build());

    public static final Type<MyEntityEntity> $TYPE = new TypeBuilder<MyEntityEntity>(MyEntityEntity.class, "MyEntity")
    .setBaseType(MyEntity.class)
    .setCacheable(true)
    .setImmutable(false)
    .setReadOnly(false)
    .setStateless(false)
    .setView(false)
    .setFactory(new Supplier<MyEntityEntity>() {
        @Override
        public MyEntityEntity get() {
            return new MyEntityEntity();
        }
    })
    .setProxyProvider(new Function<MyEntityEntity, EntityProxy<MyEntityEntity>>() {
        @Override
        public EntityProxy<MyEntityEntity> apply(MyEntityEntity entity) {
            return entity.$proxy;
        }
    })
    .addAttribute(ID)
    .build();

    private PropertyState $id_state;

    private final transient EntityProxy<MyEntityEntity> $proxy = new EntityProxy<MyEntityEntity>(this, $TYPE);

    public MyEntityEntity() {
    }

    @Override
    public Integer getId() {
        return $proxy.get(ID);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof MyEntityEntity && ((MyEntityEntity)obj).$proxy.equals(this.$proxy);
    }

    @Override
    public int hashCode() {
        return $proxy.hashCode();
    }

    @Override
    public String toString() {
        return $proxy.toString();
    }
}
```

btw. for the generated entity classes above i used ```apply plugin: 'kotlin-kapt'``` in my build.gradle but even with the 'normal' ```kapt { generateStubs = true }``` wont't work. For the abstract class with the open id property it generates:

```kotlin
public class MyEntityEntity extends MyEntity implements Persistable {
    public static final Type<MyEntityEntity> $TYPE = new TypeBuilder<MyEntityEntity>(MyEntityEntity.class, "MyEntity")
    .setBaseType(MyEntity.class)
    .setCacheable(true)
    .setImmutable(false)
    .setReadOnly(false)
    .setStateless(false)
    .setView(false)
    .setFactory(new Supplier<MyEntityEntity>() {
        @Override
        public MyEntityEntity get() {
            return new MyEntityEntity();
        }
    })
    .setProxyProvider(new Function<MyEntityEntity, EntityProxy<MyEntityEntity>>() {
        @Override
        public EntityProxy<MyEntityEntity> apply(MyEntityEntity entity) {
            return entity.$proxy;
        }
    })
    .build();

    private final transient EntityProxy<MyEntityEntity> $proxy = new EntityProxy<MyEntityEntity>(this, $TYPE);

    public MyEntityEntity() {
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof MyEntityEntity && ((MyEntityEntity)obj).$proxy.equals(this.$proxy);
    }

    @Override
    public int hashCode() {
        return $proxy.hashCode();
    }

    @Override
    public String toString() {
        return $proxy.toString();
    }
}
```

The id property is not even mentioned.

**Comment by npurushe** [comment]
This should be addressed by this PR #464 for the next update

**Comment by polmabri** [comment]
Unfortunately PR #464 don't solve this problem - i tried the latest master commit ```f58329854a``` and got this behavior:

now works:
```kotlin
@Entity
abstract class Timer : Persistable {

    @get:Key
    @get:Generated
    abstract val id: Int
}
```

produces 'TimerEntity is not abstract and does not override abstract method getMillis() in Timer' error:
```kotlin
@Entity
abstract class Timer : Persistable {

    @get:Key
    @get:Generated
    abstract val id: Int

    abstract val millis: Long
}
```

generated entity:
```kotlin
public class TimerEntity extends Timer implements Persistable {
    public static final AttributeDelegate<TimerEntity, Integer> ID = new AttributeDelegate(
    new AttributeBuilder<TimerEntity, Integer>("id", int.class)
    .setProperty(new IntProperty<TimerEntity>() {
        @Override
        public Integer get(TimerEntity entity) {
            return entity.id;
        }

        @Override
        public void set(TimerEntity entity, Integer value) {
            entity.id = value;
        }

        @Override
        public int getInt(TimerEntity entity) {
            return entity.id;
        }

        @Override
        public void setInt(TimerEntity entity, int value) {
            entity.id = value;
        }
    })
    .setPropertyName("getId")
    .setPropertyState(new Property<TimerEntity, PropertyState>() {
        @Override
        public PropertyState get(TimerEntity entity) {
            return entity.$id_state;
        }

        @Override
        public void set(TimerEntity entity, PropertyState value) {
            entity.$id_state = value;
        }
    })
    .setKey(true)
    .setGenerated(true)
    .setLazy(false)
    .setNullable(false)
    .setUnique(false)
    .build());

    public static final Type<TimerEntity> $TYPE = new TypeBuilder<TimerEntity>(TimerEntity.class, "Timer")
    .setBaseType(Timer.class)
    .setCacheable(true)
    .setImmutable(false)
    .setReadOnly(false)
    .setStateless(false)
    .setView(false)
    .setFactory(new Supplier<TimerEntity>() {
        @Override
        public TimerEntity get() {
            return new TimerEntity();
        }
    })
    .setProxyProvider(new Function<TimerEntity, EntityProxy<TimerEntity>>() {
        @Override
        public EntityProxy<TimerEntity> apply(TimerEntity entity) {
            return entity.$proxy;
        }
    })
    .addAttribute(ID)
    .build();

    private PropertyState $id_state;

    private int id;

    private final transient EntityProxy<TimerEntity> $proxy = new EntityProxy<TimerEntity>(this, $TYPE);

    public TimerEntity() {
    }

    @Override
    public int getId() {
        return $proxy.get(ID);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof TimerEntity && ((TimerEntity)obj).$proxy.equals(this.$proxy);
    }

    @Override
    public int hashCode() {
        return $proxy.hashCode();
    }

    @Override
    public String toString() {
        return $proxy.toString();
    }
}
```

**Comment by npurushe** [comment]
This fix is in 1.2.1
